routing/control_tower.go

```diff
@@ -131,6 +143,8 @@ func (p *controlTower) RegisterAttempt(paymentHash lntypes.Hash,
 	attempt *channeldb.HTLCAttemptInfo) error {
 
 	p.paymentsMtx.Lock(paymentHash)
+	p.newSubscriptionMtx.RLock()
```

Comment:
For the new subscription to `all` payments you only want to make sure that there are no updates while creating the new subscription, so you don't miss any. Ideally you would just lock the paymentsMtx when creating the 'all' subscription, but because that is hash-based, you can't lock all payments with that. That's why you need an additional mutex that can lock all payments at once. 

The payment mutex makes sure the updates for each individual payment occur in the order they are submitted to the database. So you're taking advantage of that locking in the 'all' subscription too.

Reply:
You don't want to `Lock` on a single mutex, because that would eliminate concurrent payments.

So the resulting stream to the user will have updates per payment that will be transmitted in the order they are stored in the database. But for different payments the order may vary. So the state in the database could be like this:
1) attempt on payment 1
2) attempt on payment 2

And in the update stream the user could see this:
1) attempt on payment 2
2) attempt on payment 1